File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/activity/CacheActivity.java
package com.wa2c.android.medoly.plugin.action.lrclyrics.activity;

import android.app.ActionBar;
import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;

import com.wa2c.android.medoly.plugin.action.lrclyrics.R;
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCache;
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCacheHelper;
import com.wa2c.android.medoly.plugin.action.lrclyrics.dialog.CacheDialogFragment;
import com.wa2c.android.medoly.plugin.action.lrclyrics.dialog.ConfirmDialogFragment;
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.AppUtils;
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.Logger;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Background;
import org.androidannotations.annotations.Click;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.Extra;
import org.androidannotations.annotations.ItemClick;
import org.androidannotations.annotations.OptionsItem;
import org.androidannotations.annotations.OptionsMenu;
import org.androidannotations.annotations.UiThread;
import org.androidannotations.annotations.ViewById;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;

@EActivity(R.layout.activity_cache)
@OptionsMenu(R.menu.activity_cache)
public class CacheActivity extends Activity {

    private static final int REQUEST_CODE_SAVE_FILE = 1;
    public static final String INTENT_SEARCH_TITLE = "INTENT_SEARCH_TITLE";
    public static final String INTENT_SEARCH_ARTIST = "INTENT_SEARCH_ARTIST";

    /** Search list adapter. */
    private CacheAdapter cacheAdapter;
    /** Search cache helper. */
    private SearchCacheHelper searchCacheHelper;
    /** SearchCache list. */
    private List<SearchCache> cacheList = new ArrayList<>();

    @Extra(INTENT_SEARCH_TITLE)
    String intentSearchTitle;
    @Extra(INTENT_SEARCH_ARTIST)
    String intentSearchArtist;

    @ViewById
    EditText cacheTitleEditText;
    @ViewById
    EditText cacheArtistEditText;
    @ViewById
    ListView cacheListView;

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
    }

    @AfterViews
    void afterViews() {
        ActionBar actionBar = getActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowTitleEnabled(true);
        }

        searchCacheHelper = new SearchCacheHelper(this);
        cacheAdapter = new CacheAdapter(this);
        cacheListView.setAdapter(cacheAdapter);

        cacheTitleEditText.setText(intentSearchTitle);
        cacheArtistEditText.setText(intentSearchArtist);
    }

    @OptionsItem(android.R.id.home)
    void menuHomeClick() {
        startActivity(new Intent(this, MainActivity_.class));
    }

    @OptionsItem(R.id.menu_cache_delete)
    void menuDeleteClick() {
        if (cacheAdapter.getCheckedSet().size() == 0) {
            AppUtils.showToast(this, R.string.error_delete_cache_check);
            return;
        }

        ConfirmDialogFragment dialog = ConfirmDialogFragment.newInstance(
                getString(R.string.message_dialog_cache_delete),
                getString(R.string.label_confirmation),
                getString(R.string.label_dialog_cache_delete),
                null,
                getString(android.R.string.cancel)
        );
        dialog.setClickListener(new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                if (which == DialogInterface.BUTTON_POSITIVE) {
                    deleteCache(cacheAdapter.getCheckedSet());
                }
            }
        });
        dialog.show(this);
    }

    @OptionsItem(R.id.menu_cache_open_search)
    void menuOpenCacheClick() {
        Intent intent = new Intent(this, SearchActivity_.class);
        intent.putExtra(SearchActivity.INTENT_SEARCH_TITLE, cacheTitleEditText.getText().toString());
        intent.putExtra(SearchActivity.INTENT_SEARCH_ARTIST, cacheArtistEditText.getText().toString());
        startActivity(intent);
    }

    @Background
    void deleteCache(Collection<SearchCache> caches) {
        if (searchCacheHelper.delete(cacheAdapter.getCheckedSet())) {
            cacheList.removeAll(caches);
            AppUtils.showToast(this, R.string.message_delete_cache);
            showCacheList();
        }
    }


    @Click(R.id.cacheInputClearButton)
    void cacheInputClearButtonClick() {
        cacheTitleEditText.setText(null);
        cacheArtistEditText.setText(null);
    }

    @Click(R.id.cacheInputSearchButton)
    void cacheInputSearchButtonClick(View view) {
        // Hide keyboard
        InputMethodManager inputMethodMgr = (InputMethodManager)getSystemService(INPUT_METHOD_SERVICE);
        inputMethodMgr.hideSoftInputFromWindow(view.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);

        String title = cacheTitleEditText.getText().toString();
        String artist = cacheArtistEditText.getText().toString();
        searchCache(title, artist);
    }

    // Search

    @Background
    void searchCache(String title, String artist) {
        try {
            cacheList = searchCacheHelper.search(title, artist);
            showCacheList();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @UiThread
    void showCacheList() {
        cacheAdapter.clear();
        cacheAdapter.addAll(cacheList);
        cacheAdapter.notifyDataSetChanged();
    }

    // Item

    private SearchCache currentCacheItem;

    @ItemClick(R.id.cacheListView)
    void cacheListViewItemClick(@NonNull final SearchCache item) {
        CacheDialogFragment dialog = CacheDialogFragment.newInstance(item);
        dialog.setClickListener(new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                if (which == DialogInterface.BUTTON_POSITIVE) {
                    currentCacheItem = item;
                    Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
                    intent.addCategory(Intent.CATEGORY_OPENABLE);
                    intent.setType("*/*");
                    intent.putExtra(Intent.EXTRA_TITLE, item.getResultItem().getMusicTitle() + ".lrc");
                    startActivityForResult(intent, REQUEST_CODE_SAVE_FILE);
                }
            }
        });
        dialog.show(this);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent resultData) {
        if (requestCode == REQUEST_CODE_SAVE_FILE) {
            // 歌詞のファイル保存
            if (resultCode == RESULT_OK) {
                Uri uri = resultData.getData();
                try (OutputStream stream = getContentResolver().openOutputStream(uri);
                    BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(stream,  "UTF-8"))) {
                    writer.write(currentCacheItem.getResultItem().getLyrics());
                    writer.flush();
                    AppUtils.showToast(this, R.string.message_lyrics_save_succeeded);
                } catch (IOException e) {
                    Logger.e(e);
                    AppUtils.showToast(this, R.string.message_lyrics_save_failed);
                }
            }
        }

        currentCacheItem = null;

        // Hide keyboard
        if (getCurrentFocus() != null) {
            InputMethodManager inputMethodMgr = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
            inputMethodMgr.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }


    /**
     * Search result adapter.
     */
    private static class CacheAdapter extends ArrayAdapter<SearchCache> {

        CacheAdapter(Context context) {
            super(context, R.layout.layout_search_item);
        }

        private HashSet<SearchCache> checkedSet = new HashSet<>();

        /**
         * Get checked item positions.
         * @return Checked position.
         */
        HashSet<SearchCache> getCheckedSet() {
            return checkedSet;
        }

        @NonNull
        @Override
        public View getView(final int position, View convertView, @NonNull final ViewGroup parent) {
            // view
            final ListItemViewHolder holder;
            if (convertView == null) {
                final View view = View.inflate(parent.getContext(), R.layout.layout_cache_item, null);
                holder = new ListItemViewHolder();
                holder.checkBox = (CheckBox)view.findViewById(R.id.cacheCheckBox);
                holder.titleTextView  = (TextView)view.findViewById(R.id.cacheItemTitleTextView);
                holder.artistTextView = (TextView)view.findViewById(R.id.cacheItemArtistTextView);
                holder.fromTextView  = (TextView)view.findViewById(R.id.cacheItemFromTextView);
                holder.fileTextView = (TextView)view.findViewById(R.id.cacheItemFileTextView);
                holder.langTextView = (TextView)view.findViewById(R.id.cacheItemLangTextView);
                view.setTag(holder);
                convertView = view;
            } else {
                holder = (ListItemViewHolder) convertView.getTag();
            }

            // data
            final SearchCache item = getItem(position);
            holder.checkBox.setChecked(checkedSet.contains(item));
            holder.titleTextView.setText(getContext().getString(R.string.label_cache_item_title, AppUtils.coalesce(item.title)));
            holder.artistTextView.setText(getContext().getString(R.string.label_cache_item_artist, AppUtils.coalesce(item.artist)));
            holder.fromTextView.setText(getContext().getString(R.string.label_cache_item_from, AppUtils.coalesce(item.from)));
            holder.fileTextView.setText(getContext().getString(R.string.label_cache_item_file, AppUtils.coalesce(item.file_name)));
            holder.langTextView.setText(getContext().getString(R.string.label_cache_item_lang, AppUtils.coalesce(item.language)));

            // event
            holder.checkBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    if (isChecked)
                        checkedSet.add(item);
                    else
                        checkedSet.remove(item);
                }
            });

            return convertView;
        }

        @Override
        public void notifyDataSetChanged() {
            super.notifyDataSetChanged();
            checkedSet.clear();
        }

    }

    /** List item view holder. */
    private static class ListItemViewHolder {
        CheckBox checkBox;
        TextView titleTextView;
        TextView artistTextView;
        TextView fromTextView;
        TextView fileTextView;
        TextView langTextView;
    }


}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/activity/CacheActivity.kt
package com.wa2c.android.medoly.plugin.action.lrclyrics.activity

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.wa2c.android.medoly.plugin.action.lrclyrics.R
import com.wa2c.android.medoly.plugin.action.lrclyrics.databinding.ActivityCacheBinding
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.AppDatabase
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCache
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCacheDao
import com.wa2c.android.medoly.plugin.action.lrclyrics.dialog.CacheDialogFragment
import com.wa2c.android.medoly.plugin.action.lrclyrics.dialog.ConfirmDialogFragment
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.AppUtils
import kotlinx.android.synthetic.main.layout_cache_item.view.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.*

/**
 * Cache activity
 */
class CacheActivity : AppCompatActivity() {

    /** Binding. */
    private lateinit var binding: ActivityCacheBinding

    /** Search list adapter.  */
    private lateinit var cacheAdapter: CacheAdapter
    /** Search cache DAO. */
    private lateinit var dao: SearchCacheDao
    /** Current cache item. */
    private var currentCacheItem: SearchCache? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_cache)

        // Action Bar
        supportActionBar?.let {
            it.setDisplayShowHomeEnabled(true)
            it.setDisplayHomeAsUpEnabled(true)
            it.setDisplayShowTitleEnabled(true)
        }

        dao = AppDatabase.buildDb(this).getSearchCacheDao()
        cacheAdapter = CacheAdapter(this)
        binding.cacheListView.adapter = cacheAdapter

        binding.cacheTitleEditText.setText(intent.getStringExtra(INTENT_SEARCH_TITLE))
        binding.cacheArtistEditText.setText(intent.getStringExtra(INTENT_SEARCH_ARTIST))

        binding.cacheInputClearButton.setOnClickListener {
            binding.cacheTitleEditText.text = null
            binding.cacheArtistEditText.text = null
        }

        binding.cacheInputSearchButton.setOnClickListener {
            // Hide keyboard
            val inputMethodMgr = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodMgr.hideSoftInputFromWindow(it.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)

            val title = binding.cacheTitleEditText.text.toString()
            val artist = binding.cacheArtistEditText.text.toString()

            searchCache(title, artist)
            binding.cacheTitleEditText.tag = title
            binding.cacheArtistEditText.tag = artist
        }

        binding.cacheListView.setOnItemClickListener { _, _, position, _ ->
            val item  = cacheAdapter.getItem(position)!!
            val dialog = CacheDialogFragment.newInstance(item)

            dialog.clickListener = { _, which, _ ->
                when (which) {
                    // Save
                    DialogInterface.BUTTON_POSITIVE -> {
                        currentCacheItem = item
                        AppUtils.saveFile(this@CacheActivity, item.title, item.artist)
                    }
                    // Re-search
                    DialogInterface.BUTTON_NEGATIVE -> {
                        currentCacheItem = item
                        val intent = Intent(this@CacheActivity, SearchActivity::class.java)
                        intent.putExtra(SearchActivity.INTENT_SEARCH_TITLE, item.title)
                        intent.putExtra(SearchActivity.INTENT_SEARCH_ARTIST, item.artist)
                        startActivity(intent)
                    }
                    // Delete lyrics
                    CacheDialogFragment.DIALOG_RESULT_DELETE_LYRICS -> searchCache(binding.cacheTitleEditText.tag as String, binding.cacheArtistEditText.tag as String)
                    // Delete cache
                    CacheDialogFragment.DIALOG_RESULT_DELETE_CACHE -> searchCache(binding.cacheTitleEditText.tag as String,binding. cacheArtistEditText.tag as String)
                }
            }
            dialog.show(this)
        }

    }

    /**
     * On New Intent
     */
    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)

        binding.cacheListView.adapter = cacheAdapter
        binding.cacheTitleEditText.setText(intent.getStringExtra(INTENT_SEARCH_TITLE))
        binding.cacheArtistEditText.setText(intent.getStringExtra(INTENT_SEARCH_ARTIST))
    }

    /**
     * On create options menu.
     */
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.activity_cache, menu)
        if (menu != null) {
            for (i in 0 until menu.size()) {
                menu.getItem(i).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
            }
        }
        return true
    }

    /**
     * On options item selected.
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                startActivity(Intent(this, MainActivity::class.java))
                return true
            }
            R.id.menu_cache_delete -> {
                if (cacheAdapter.checkedSet.size == 0) {
                    AppUtils.showToast(this, R.string.error_delete_cache_check)
                    return true
                }

                val dialog = ConfirmDialogFragment.newInstance(
                        getString(R.string.message_dialog_cache_delete),
                        getString(R.string.label_confirmation),
                        getString(R.string.label_dialog_cache_delete),
                        null,
                        getString(android.R.string.cancel)
                )

                dialog.clickListener = { _, which, _ ->
                    if (which == DialogInterface.BUTTON_POSITIVE) {
                        GlobalScope.launch(Dispatchers.Main) {
                            val result = async(Dispatchers.Default) {
                                return@async (dao.deleteIn(cacheAdapter.checkedSet.map{ r -> r._id }.toList()) > 0)
                            }
                            if (result.await()) {
                                cacheAdapter.removeCheckedItem()
                                AppUtils.showToast(this@CacheActivity, R.string.message_delete_cache)
                            }
                        }
                    }
                }
                dialog.show(this)
            }
            R.id.menu_cache_open_search -> {
                val intent = Intent(this, SearchActivity::class.java)
                intent.putExtra(SearchActivity.INTENT_SEARCH_TITLE, binding.cacheTitleEditText.text.toString())
                intent.putExtra(SearchActivity.INTENT_SEARCH_ARTIST, binding.cacheArtistEditText.text.toString())
                startActivity(intent)
            }
        }

        return super.onOptionsItemSelected(item)
    }

    /**
     * On activity result
     */
    public override fun onActivityResult(requestCode: Int, resultCode: Int, resultData: Intent?) {
        if (requestCode == AppUtils.REQUEST_CODE_SAVE_FILE && resultCode == RESULT_OK) {
            // Save to lyrics file
            try {
                val uri = resultData?.data!!
                contentResolver.openOutputStream(uri)!!.bufferedWriter(Charsets.UTF_8).use {
                    it.write(currentCacheItem!!.makeResultItem()!!.lyrics)
                }
                AppUtils.showToast(this, R.string.message_lyrics_save_succeeded)
            } catch (e: Exception) {
                Timber.e(e)
                AppUtils.showToast(this, R.string.message_lyrics_save_failed)
            }
        }

        currentCacheItem = null

        // Hide keyboard
        if (currentFocus != null) {
            val inputMethodMgr = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodMgr.hideSoftInputFromWindow(currentFocus!!.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }

    /**
     * Search cache
     */
    private fun searchCache(title: String, artist: String) {
        GlobalScope.launch(Dispatchers.Main) {
            val result = async(Dispatchers.Default) {
                return@async dao.search(title, artist)
            }
            cacheAdapter.setList(result.await().toMutableList())
        }
    }



    /**
     * Search result adapter.
     */
    private class CacheAdapter internal constructor(context: Context) : ArrayAdapter<SearchCache>(context, R.layout.layout_search_item) {
        /** Checked item. */
        val checkedSet = HashSet<SearchCache>()
        /** SearchCache list.  */
        private var cacheList: MutableList<SearchCache> = ArrayList()

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var itemView = convertView
            // itemView
            val holder: ListItemViewHolder
            if (itemView == null) {
                holder = ListItemViewHolder(parent.context)
                itemView = holder.itemView
            } else {
                holder = itemView.tag as ListItemViewHolder
            }

            holder.bind(getItem(position)!!, checkedSet)

            return itemView
        }

        fun setList(list: MutableList<SearchCache>) {
            cacheList = list
            showCacheList()
        }

        fun removeCheckedItem() {
            cacheList.removeAll(checkedSet)
            showCacheList()
        }

        fun showCacheList() {
            clear()
            addAll(cacheList)
            notifyDataSetChanged()
        }

        override fun notifyDataSetChanged() {
            super.notifyDataSetChanged()
            checkedSet.clear()
        }

        /** List item view holder.  */
        private class ListItemViewHolder(val context: Context) {
            val itemView = View.inflate(context, R.layout.layout_cache_item, null)!!
            init {
                itemView.tag = this
            }

            fun bind(item: SearchCache, checkedSet: HashSet<SearchCache>) {
                itemView.cacheItemCheckBox.isChecked = checkedSet.contains(item)
                itemView.cacheItemTitleTextView.text = context.getString(R.string.label_cache_item_title, AppUtils.coalesce(item.title))
                itemView.cacheItemArtistTextView.text = context.getString(R.string.label_cache_item_artist, AppUtils.coalesce(item.artist))
                itemView.cacheItemFromTextView.text = context.getString(R.string.label_cache_item_from, AppUtils.coalesce(item.from))
                itemView.cacheItemFileTextView.text = context.getString(R.string.label_cache_item_file, AppUtils.coalesce(item.file_name))
                itemView.cacheItemLangTextView.text = context.getString(R.string.label_cache_item_lang, AppUtils.coalesce(item.language))
                if (item.has_lyrics == true) {
                    itemView.cacheItemHasLyricsImageView.visibility = View.VISIBLE
                } else {
                    itemView.cacheItemHasLyricsImageView.visibility = View.GONE
                }

                // event
                itemView.cacheItemCheckBox.setOnCheckedChangeListener { _, isChecked ->
                    if (isChecked)
                        checkedSet.add(item)
                    else
                        checkedSet.remove(item)
                }
            }
        }
    }



    companion object {
        const val INTENT_SEARCH_TITLE = "INTENT_SEARCH_TITLE"
        const val INTENT_SEARCH_ARTIST = "INTENT_SEARCH_ARTIST"
    }

}
